src/LanguageServer/Impl/Indexing/SymbolIndex.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Python.Analysis.Documents;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.IO;
+using Microsoft.Python.Parsing;
+
+namespace Microsoft.Python.LanguageServer.Indexing {
+    internal sealed class SymbolIndex : ISymbolIndex {
+        private readonly ConcurrentDictionary<string, IMostRecentDocumentSymbols> _index;
+        private readonly IFileSystem _fileSystem;
+        private readonly PythonLanguageVersion _version;
+
+        public SymbolIndex(IFileSystem fileSystem, PythonLanguageVersion version) {
+            _fileSystem = fileSystem;
+            _version = version;
+
+            var comparer = PathEqualityComparer.Instance;
+            _index = new ConcurrentDictionary<string, IMostRecentDocumentSymbols>(comparer);
+        }
+
+        public Task<IReadOnlyList<HierarchicalSymbol>> HierarchicalDocumentSymbolsAsync(string path, CancellationToken ct = default) {
+            if (_index.TryGetValue(path, out var mostRecentSymbols)) {
+                return mostRecentSymbols.GetSymbolsAsync(ct);
+            } else {
+                return Task.FromResult<IReadOnlyList<HierarchicalSymbol>>(new List<HierarchicalSymbol>());
+            }
+        }
+
+        public async Task<IReadOnlyList<FlatSymbol>> WorkspaceSymbolsAsync(string query, int maxLength, CancellationToken ct = default) {
+            var tasks = _index
+                .Select(kvp => WorkspaceSymbolsQueryAsync(kvp.Key, query, kvp.Value, ct))
+                .ToArray();
```

Comment:
Won't this blow up? This to me reads as:

1. Spawn as many tasks as there are files in the workspace (uh oh).
2. Each task collects its own read only list of flat symbols (loads of allocations).
3. Then limit once each one is done (too late to ward off any of the previous things).

For small test cases, this works, but use it on any reasonably large project and this is going to hang until cancelled at best.

Reply:
Yep, we need a `SemaphoreSlim` here to control amount of concurrently running tasks (`Parallel.For` won't work because of async).

Reply:
I think that it's even past that; sending an empty string as the query will return every symbol that has been indexed, and this code is going to collect them. In previous versions (before the async stuff), the enumerable would be limited as it is doing its modified BFS. Now it's a two step process where all these tasks are spawned, waited for every file, then limited. I think it's going to have to be done in another way to stop the query early at the provided limit.

Reply:
`maxLength` is still too big, 1000 tasks will paralyze the system.